src/trace/log_backend_probe.c

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/zephyr.h>
+
+#include <sof/audio/buffer.h>
+#include <sof/ipc/topology.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/probe/probe.h>
+
+/*
+ * A lock is needed as log_process() and log_panic() have no internal locks
+ * to prevent concurrency. Meaning if log_process is called after
+ * log_panic was called previously, log_process may happen from another
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock probe_lock;
+
+static uint32_t probe_log_format_current = CONFIG_LOG_BACKEND_SOF_PROBE_OUTPUT;
+
+#define LOG_BUF_SIZE 80
+static uint8_t log_buf[LOG_BUF_SIZE];
+
+static probe_logging_hook_t probe_hook;
+
+static int probe_char_out(uint8_t *data, size_t length, void *ctx)
+{
+	if (probe_hook)
+		probe_hook(data, length);
+
+	return length;
+}
+
+LOG_OUTPUT_DEFINE(log_output_adsp_probe, probe_char_out, log_buf, sizeof(log_buf));
+
+static uint32_t format_flags(void)
+{
+	uint32_t flags = LOG_OUTPUT_FLAG_LEVEL | LOG_OUTPUT_FLAG_TIMESTAMP;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_FORMAT_TIMESTAMP))
+		flags |= LOG_OUTPUT_FLAG_FORMAT_TIMESTAMP;
+
+	return flags;
+}
+
+static void probe_log_panic(struct log_backend const *const backend)
+{
+	k_spinlock_key_t key = k_spin_lock(&probe_lock);
+
+	log_backend_std_panic(&log_output_adsp_probe);
+
+	k_spin_unlock(&probe_lock, key);
+}
+
+static inline void probe_log_dropped(const struct log_backend *const backend,
```

Comment:
this cannot be `inline` - its pointer is used

Reply:
Oops, similar pattern in Zephyr side backends. FYI @teburd  Will fix here.